File: src/ui_tree.rs
use std::collections::HashMap;

use anyhow::{anyhow, Result};
use quick_xml::events::{BytesStart, Event};
use quick_xml::Reader;

/// A rectangle in device/screen pixel coordinates: [left, top] -> [right, bottom].
#[derive(Debug, Clone, Copy)]
pub struct Bounds {
    pub left: i32,
    pub top: i32,
    pub right: i32,
    pub bottom: i32,
}

impl Bounds {
    pub fn width(&self) -> i32 {
        self.right - self.left
    }
    pub fn height(&self) -> i32 {
        self.bottom - self.top
    }
    pub fn contains(&self, x: i32, y: i32) -> bool {
        x >= self.left && x <= self.right && y >= self.top && y <= self.bottom
    }
}

/// A node in the UI hierarchy, mirroring a `<node>` element from uiautomator.
#[derive(Debug, Clone)]
pub struct Node {
    pub id: usize,
    pub attrs: HashMap<String, String>,
    pub bounds: Option<Bounds>,
    pub children: Vec<Node>,
}

impl Node {
    fn new() -> Self {
        Node {
            id: 0,
            attrs: HashMap::new(),
            bounds: None,
            children: Vec::new(),
        }
    }

    /// Total number of nodes in this subtree (including self).
    pub fn count(&self) -> usize {
        1 + self.children.iter().map(|c| c.count()).sum::<usize>()
    }

    /// Find a node by its stable id (assigned at parse time).
    pub fn find(&self, id: usize) -> Option<&Node> {
        if self.id == id {
            return Some(self);
        }
        for c in &self.children {
            if let Some(n) = c.find(id) {
                return Some(n);
            }
        }
        None
    }

    /// Return the id of the deepest (smallest-area) node whose bounds contain (x, y),
    /// simulating uiautomatorviewer's "pick the innermost element" behavior.
    pub fn hit_test(&self, x: i32, y: i32) -> Option<usize> {
        fn rec(node: &Node, x: i32, y: i32, best: &mut Option<(usize, i64)>) {
            if let Some(b) = &node.bounds {
                if b.contains(x, y) {
                    let area = b.width() as i64 * b.height() as i64;
                    if best.is_none() || area < best.unwrap().1 {
                        *best = Some((node.id, area));
                    }
                }
            }
            for c in &node.children {
                rec(c, x, y, best);
            }
        }
        let mut best = None;
        rec(self, x, y, &mut best);
        best.map(|(id, _)| id)
    }

    /// True if this node (or any descendant) matches the search query in any attribute.
    pub fn subtree_matches(&self, q: &str) -> bool {
        let q = q.to_lowercase();
        self.attrs.values().any(|v| v.to_lowercase().contains(&q))
            || self.children.iter().any(|c| c.subtree_matches(&q))
    }
}

fn parse_bounds(s: &str) -> Option<Bounds> {
    let nums: Vec<i32> = s
        .split(|c: char| c == '[' || c == ']' || c == ',')
        .filter_map(|p| p.trim().parse::<i32>().ok())
        .collect();
    if nums.len() == 4 {
        Some(Bounds {
            left: nums[0],
            top: nums[1],
            right: nums[2],
            bottom: nums[3],
        })
    } else {
        None
    }
}

fn build_node(e: &BytesStart, counter: &mut usize) -> Result<Node> {
    let mut node = Node::new();
    node.id = *counter;
    *counter += 1;

    for a in e.attributes() {
        let a = a?;
        // uiautomator XML attributes carry no namespace prefix, so the full key
        // (via AsRef<[u8]>) is identical to the local name.
        let key = String::from_utf8_lossy(a.key.as_ref()).into_owned();
        let val: String = a.unescape_value()?.into_owned();
        node.attrs.insert(key, val);
    }
    node.bounds = parse_bounds(node.attrs.get("bounds").map(|s| s.as_str()).unwrap_or(""));
    Ok(node)
}

/// Parse a uiautomator XML dump into a `Node` tree rooted at `<hierarchy>`.
pub fn parse(xml: &str) -> Result<Node> {
    let mut reader = Reader::from_str(xml);

    let mut stack: Vec<Node> = Vec::new();
    let mut root: Option<Node> = None;
    let mut counter: usize = 0;
    let mut buf = Vec::new();

    loop {
        buf.clear();
        match reader.read_event_into(&mut buf) {
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                let node = build_node(&e, &mut counter)?;
                stack.push(node);
            }
            Ok(Event::Empty(e)) => {
                let node = build_node(&e, &mut counter)?;
                if let Some(parent) = stack.last_mut() {
                    parent.children.push(node);
                }
            }
            Ok(Event::End(_)) => {
                let node = stack
                    .pop()
                    .ok_or_else(|| anyhow!("XML 结构异常：多余的结束标签"))?;
                match stack.last_mut() {
                    Some(parent) => parent.children.push(node),
                    None => root = Some(node),
                }
            }
            Ok(_) => {}
            Err(e) => {
                return Err(anyhow!(
                    "XML 解析错误 (位置 {}): {}",
                    reader.buffer_position(),
                    e
                ))
            }
        }
    }

    root.ok_or_else(|| anyhow!("未找到 hierarchy 根节点"))
}

File: web/backend/uitree.py
"""Python port of the Rust `ui_tree.rs`.

Parses a uiautomator XML hierarchy into a nested `Node` tree with stable ids,
and provides the same two helpers used by the desktop app:
  - `hit_test(x, y)` -> innermost (smallest-area) node containing a point
  - `subtree_matches(q)` -> any attr contains the search query (for filtering)
"""

from __future__ import annotations

import re
import xml.etree.ElementTree as ET
from typing import Optional

_BOUNDS_RE = re.compile(r"[\[\],]")


class Bounds:
    __slots__ = ("left", "top", "right", "bottom")

    def __init__(self, left: int, top: int, right: int, bottom: int):
        self.left = left
        self.top = top
        self.right = right
        self.bottom = bottom

    @property
    def width(self) -> int:
        return self.right - self.left

    @property
    def height(self) -> int:
        return self.bottom - self.top

    def contains(self, x: int, y: int) -> bool:
        return self.left <= x <= self.right and self.top <= y <= self.bottom


class Node:
    __slots__ = ("id", "attrs", "bounds", "children")

    def __init__(self, node_id: int, attrs: dict[str, str], bounds: Optional[Bounds]):
        self.id = node_id
        self.attrs = attrs
        self.bounds = bounds
        self.children: list["Node"] = []

    def count(self) -> int:
        return 1 + sum(c.count() for c in self.children)

    def find(self, node_id: int) -> Optional["Node"]:
        if self.id == node_id:
            return self
        for c in self.children:
            found = c.find(node_id)
            if found:
                return found
        return None

    def hit_test(self, x: int, y: int) -> Optional[int]:
        best: Optional[tuple[int, int]] = None  # (id, area)

        def rec(node: "Node"):
            nonlocal best
            if node.bounds and node.bounds.contains(x, y):
                area = node.bounds.width * node.bounds.height
                if best is None or area < best[1]:
                    best = (node.id, area)
            for c in node.children:
                rec(c)

        rec(self)
        return best[0] if best else None

    def subtree_matches(self, q: str) -> bool:
        q = q.lower()
        if any(q in v.lower() for v in self.attrs.values()):
            return True
        return any(c.subtree_matches(q) for c in self.children)

    def to_dict(self) -> dict:
        b = self.bounds
        return {
            "id": self.id,
            "attrs": self.attrs,
            "bounds": {
                "left": b.left,
                "top": b.top,
                "right": b.right,
                "bottom": b.bottom,
            }
            if b
            else None,
            "children": [c.to_dict() for c in self.children],
        }


def _parse_bounds(s: str) -> Optional[Bounds]:
    nums = [int(p) for p in _BOUNDS_RE.split(s) if p.strip().isdigit()]
    if len(nums) == 4:
        return Bounds(nums[0], nums[1], nums[2], nums[3])
    return None


def parse(xml: str) -> Node:
    """Parse a uiautomator XML dump into a `Node` tree rooted at <hierarchy>."""
    root_el = ET.fromstring(xml)
    counter = 0

    def build(el) -> Node:
        nonlocal counter
        node_id = counter
        counter += 1
        attrs = {k: (v if v is not None else "") for k, v in el.attrib.items()}
        bounds = _parse_bounds(attrs.get("bounds", ""))
        node = Node(node_id, attrs, bounds)
        for child in el:
            node.children.append(build(child))
        return node

    return build(root_el)
